vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+            }
+            worker.process_main().await
+        };
+
+        match stop.until_stopped(fut).await? {
+            Ok(_) => {}
+            Err(err) => tracing::error!(error = err.as_error(), "storvsc run error"),
+        }
+        Ok(())
+    }
+}
+
+impl<T: 'static + Send + Sync + RingMem> InspectTask<Storvsc<T>> for StorvscState {
+    fn inspect(&self, req: inspect::Request<'_>, worker: Option<&Storvsc<T>>) {
+        if let Some(worker) = worker {
+            let mut resp = req.respond();
+            resp.field("has_negotiated", worker.has_negotiated);
+        }
+    }
+}
+
+impl<T: 'static + Send + Sync + RingMem> Storvsc<T> {
+    pub(crate) fn new(
+        channel: RawAsyncChannel<T>,
+        version: storvsp_protocol::ProtocolVersion,
+        new_request_receiver: Receiver<StorvscRequest>,
+    ) -> Result<Self, StorvscError> {
+        let queue = Queue::new(channel).map_err(StorvscError::Queue)?;
+
+        Ok(Self {
+            inner: StorvscInner {
+                next_transaction_id: AtomicU64::new(1),
+                new_request_receiver,
+                transactions: Mutex::new(HashMap::new()),
+            },
+            version,
+            queue,
+            num_sub_channels: None,
+            has_negotiated: false,
+        })
+    }
+}
+
+impl<T: Send + Sync + RingMem> Storvsc<T> {
+    async fn negotiate(&mut self) -> Result<(), StorvscError> {
+        // Negotiate protocol with storvsp instance on the other end of VMBus
+        // Step 1: BEGIN_INITIALIZATION
+        self.inner
+            .send_packet_and_expect_completion(
+                &mut self.queue,
+                storvsp_protocol::Operation::BEGIN_INITIALIZATION,
+                &(),
+            )
+            .await?;
+
+        // Step 2: QUERY_PROTOCOL_VERSION - request latest version
+        self.inner
+            .send_packet_and_expect_completion(
+                &mut self.queue,
+                storvsp_protocol::Operation::QUERY_PROTOCOL_VERSION,
+                &self.version,
+            )
+            .await
+            .map_err(|err| match err {
+                StorvscError::PacketError(PacketError::UnexpectedStatus(
+                    storvsp_protocol::NtStatus::INVALID_DEVICE_STATE,
+                )) => StorvscError::UnsupportedProtocolVersion,
+                _ => err,
+            })?;
+
+        // Step 3: QUERY_PROPERTIES
+        let properties_packet = self
+            .inner
+            .send_packet_and_expect_completion(
+                &mut self.queue,
+                storvsp_protocol::Operation::QUERY_PROPERTIES,
+                &(),
+            )
+            .await?;
+        let properties = storvsp_protocol::ChannelProperties::ref_from_prefix(
+            &properties_packet.data[0..properties_packet.data_size],
+        )
+        .map_err(|_err| StorvscError::UnexpectedOperation)?
+        .0
+        .to_owned();
+
+        // Step 4: CREATE_SUB_CHANNELS (if supported)
```

Comment:
We don't really support subchannels yet, right?

Reply:
Subchannels are seemingly supported by our storvsp implementation.

Reply:
Sure but this client doesn't seem to actually do anything with them.

I also don't know if subchannels are supported by the Linux vmbus uio driver.

Reply:
(For the use case of accessing low-perf devices, it doesn't matter. But subchannels will be a big deal for high-perf devices.)

Reply:
Resolving because subchannel support was removed in a later update to this PR